docs/development.md

```diff
@@ -0,0 +1,58 @@
+# Useful documentation
+
+* [Docker](https://docs.docker.com/)
+* [Docker-compose](https://docs.docker.com/compose/)
+* [Main Travis docs](https://docs.travis-ci.com/)
+* [Travis API](https://developer.travis-ci.com/)
+
+
+# Useful patterns and commands
+##  During development of standalone Dockerfiles
+
+If something fails during the build you can determine the command where it failed and comment out the remaining of the Dockefile. Then rebuild and tag for instance
+`test_image`. You can then interactively try to follow commented out section of Dockefile, by starting container interactively
+```
+   docker run -it --rm test_image /bin/bash
+```
+Usually in this way or similar it is way easier to debug errors.
+
+## During development of the test cases
+
+If something fails during running `docker-compose` due to communication problems or invalid input, the easiest way to debug is:
+
+1. Add sleep timer (e.g. `sleep 1000`) after running failed command
+2. Rerun the failed setup
+3. Attach to the failed container with e.g.
+```
+   docker-compose run -it calculix-adapter
+```
+4. You can then try to rerun command in `docker-compose.yml` and check the folders `/home/precice/Data/Input`, `/home/precice/Data/Exchange`, 
+`/home/precice/Data/Output`. Make sure the `precice-config.xml` is properly setup, input can be read and folders have proper rights.
```

Comment:
What does "proper rights" mean? I think some comments on this might be helpful.

From my perspective there are two points, where things with rights can go wrong:

1. **provide data to the container:** User `precice` needs the rights to read (and write?) the data in the mounted folder.
2. **obtain data from the container** The host (or another container) needs the rights to read (and write?) the data written by the container into the folder.

My current use case is the following: I `docker run -ti` two containers (e.g. `fenics-adapter`) that do all the computation. The host provides the case setup (e.g. `/tutorials/HT/partitioned-heat/fenics-fenics`) using `-v`. If the host does not set the user rights properly, `precice` cannot access the case and write output to the folder and aborts.

Reply:
You can achieve it something like:

```
docker build -f Dockerfile.Ubuntu1804.home -t precice-user --build-arg uid=$UID --build-arg gid=$GID . 
docker build -f Dockerfile.fenics-adapter -t fenics-adapter-user --build-arg from=precice-user
```
where GID and UID are you group id (obtain e.g. with `id -a` on Linux). Then you run the container with:
```
docker run -ti --user=precice fenics-adapter-user /bin/bash
```

I you want to run two containers: 

```
docker volume create exchange
docker run --user=precice -it -v $(pwd)/HT/partitioned-heat/fenics-fenics:/home/precice/Data/Input -v exchange:/home/precice/Data/Exchange fenics-fenics-user /bin/bash```

Reply:
Thanks! Did you already add these explanations to the documentation? As far as I understand `--build-arg uid` and `--build-arg gid` are not provided by docker, but you added them to the Dockerfiles (e.g. [here](https://github.com/precice/systemtests/blob/06fe89a4cd5af34ba8401b5c2f8631c6b1bd185a/Dockerfile.Ubuntu1604.home#L33-L34)) in order to sort out the user rights for our specific setup of docker containers.

Some questions the documentation should answer in this context:

* Why do we actually have to worry about creating a preCICE user?  
* Why not just run everything as `root` (which is often done in docker)? We, for example, need a non-root user for testing `make install` in a user folder. We rely at some places on `pip` for installing packages and you should not run `pip` as sudo (correct @fsimonis?). 
* What can we do with the `uid` and `gid` argument? See example above.

Reply:
Yes. Explanation that I wrote above should be in `docs/development.md` now. I also added some motivation behind the solution in `docs/architecture.md`.